classes/plugins/PKPPubIdPlugin.inc.php

```diff
@@ -0,0 +1,428 @@
+			$newPubId = $this->constructPubId($pubIdPrefix, $fieldValue, $contextId);
+
+			if (!$this->checkDuplicate($newPubId, $pubObject, $contextId)) {
+				$errorMsg = $this->getNotUniqueErrorMsg();
+				return false;
+			}
+		}
+		return true;
+	}
+
+	/**
+	 * Check whether the given pubId is valid.
+	 * @param $pubId string
+	 * @return boolean
+	 */
+	function validatePubId($pubId) {
+		return true; // Assume a valid ID by default;
+	}
+
+	/**
+	 * Return an array of publication object types and
+	 * the corresponding DAOs.
+	 * @return array
+	 */
+	abstract function getDAOs();
+
+	/**
+	 * Can a pub id be assigned to the object.
+	 * @param $pubObject object
+	 * 	OJS Issue, Article, Representation or SubmissionFile
+	 * @return boolean
+	 * 	false, if the pub id contains an unresolved pattern i.e. '%' or
+	 * 	if the custom suffix is empty i.e. the pub id null.
+	 */
+	function canBeAssigned($pubObject) {
+		// Has the pub id already been assigned.
+		$pubIdType = $this->getPubIdType();
+		$storedPubId = $pubObject->getStoredPubId($pubIdType);
+		if ($storedPubId) return false;
+		// Get the pub id.
+		$pubId = $this->getPubId($pubObject);
+		// Is the custom suffix empty i.e. the pub id null.
+		if (!$pubId) return false;
+		// Does the suffix contain unresolved pattern.
+		$containPatterns = strpos($pubId, '%') !== false;
+		return !$containPatterns;
+	}
+
+	/**
+	 * Add the suffix element and the public identifier
+	 * to the object.
+	 * @param $hookName string
+	 * @param $params array
+	 */
+	function getAdditionalFieldNames($hookName, $params) {
+		$fields =& $params[1];
+		$formFieldNames = $this->getFormFieldNames();
+		foreach ($formFieldNames as $formFieldName) {
+			$fields[] = $formFieldName;
+		}
+		$daoFieldNames = $this->getDAOFieldNames();
+		foreach ($daoFieldNames as $daoFieldName) {
+			$fields[] = $daoFieldName;
+		}
+		return false;
+	}
+
+	/**
+	 * Return the object type.
+	 * @param $pubObject object
+	 * @return array
+	 */
+	function getPubObjectType($pubObject) {
+		$allowedTypes = $this->getPubObjectTypes();
+		$pubObjectType = null;
+		foreach ($allowedTypes as $allowedType) {
+			if (is_a($pubObject, $allowedType)) {
+				$pubObjectType = $allowedType;
+				break;
+			}
+		}
+		if (is_null($pubObjectType)) {
+			// This must be a dev error, so bail with an assertion.
+			assert(false);
+			return null;
+		}
+		return $pubObjectType;
+	}
+
+	/**
+	 * Set and store a public identifier.
+	 * @param $pubObject object
+	 * @param $pubObjectType string As returned from self::getPubObjectType()
+	 * @param $pubId string
+	 * @return string
+	 *
+	 * This function is currently only used in the pubId import/export plugin.
+	 * After the migration of that plugin, check if this function is still needed.
+	 */
+	function setStoredPubId(&$pubObject, $pubObjectType, $pubId) {
+		$daos = $this->getDAOs();
+		$dao = $daos[$pubObjectType];
+		$dao->changePubId($pubObject->getId(), $this->getPubIdType(), $pubId);
```

Comment:
I think it would make sense to define an interface that DAOs would need to implement in order for pub ID support to be added, and add an assertion e.g. here checking to ensure that supplied DAOs do in fact implement it. Make sense?